Change G27 to allow Z raise if homing needed

classicrocker883 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description

When using `G27 P3`, this allows Z axis to be raised even if homing may be needed.  
Same for `G27 P4`, allows XY to move if Z axis is not homed.

When a blocking move is done only in the Z axis, it moves up by `NOZZLE_PARK_Z_RAISE_MIN` which is around 10mm. 

- Change `G27 P3` to allow Z raise if homing needed
- - Allow XY if Z !homed (`G27 P4`)
- Rename  "Nozzle Park" to "Park Nozzle"
<br>

---
-  ~~tweak ProUI caselight~~ moved to other **PR**
- ~~update **macros.h** so previous `#define` are above next call~~
- ~~add millis_t to **libs/duration_t.h**~~
- ~~duration_t.h/compat.h/stepper.cpp/motion.h/motion.cpp~~
<br>

unrelated but should be addressed
-  **src\lcd\extui\ftdi_eve_touch_ui\ftdi_eve_lib\compat.h** `IF_DISABLED(O,A)    TERN(O,,A)`
- - ~~before it was~~ defined as `_TERN(_ENA_1(O),,A)` which is the same as `TERN_`, which doesnt make sense.
- - it should be `IF_DISABLED(O,A)    TERN(O,,A)`, no?
- same with **module/stepper.cpp**, under `FT_MOTION` - `ftMotion_stepper()`, `ENDSTOP_INTERRUPTS_FEATURE` should enable the update, not disable??? (https://github.com/classicrocker883/MRiscoCProUI/issues/113)
<br>




<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits

<!-- What does this PR fix or improve? -->

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues
The only foreseen negative consequence of this change is if the axis is already near its max limit. This really should not be an issue, a user doesn't have to be extra careful in that regard knowing where the limit is.

But if a user just wants to raise the Z, homing isn't required, and this can be beneficial when aborting a print, when a print is complete, or just wanting to raise Z say if the head is too close to the bed, or printed part, for any reason.  
Then `G27 P3` can be used without needing to "Home"/"trusted axis"
<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

classicrocker883 committed:
change G27 to allow Z raise !homing

classicrocker883 committed:
update to unsigned, change to ushortval

classicrocker883 committed:
update with less wording

classicrocker883 committed:
update caselight_uses_brightness, add apply

classicrocker883 committed:
Update dwin.cpp, revert caselight

classicrocker883 committed:
amend duration_t.h, millis_t/with_days

classicrocker883 committed:
revert

classicrocker883 committed:
ftdi_eve compat.h, IF_DISABLED => TERN

classicrocker883 committed:
ENDSTOP_INTERRUPTS_FEATURE now enabled in FT_MOTION stepper.cpp

classicrocker883 committed:
if defined - report_real_position

classicrocker883 committed:
append G27 to include switch case

classicrocker883 committed:
arrange macros.h (intellisense)

classicrocker883 committed:
update bltouch spacing/comments

classicrocker883 committed:
update probe spacing/comments

classicrocker883 committed:
tweak bltouch spacing

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Update stepper.cpp - update anyway

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x-Feb' of https://github.com/classicrocker883/Marlin into bugfix-2.1.x-Feb

sjasonsmith commented:
I think your change to G27 is worth considering. Can you remove all the other unrelated changes from the PR, so that the G27 change can be reviewed and evaluated in isolation?

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
revert files but pause/G27.cpp

classicrocker883 commented:
done

sjasonsmith commented:
@InsanityAutomation now that the scope of this PR is very small, I'd like your opinion on these G-Code changes. The changes seem reasonable to me. Yes, with the Z-raise it may be possible to crash to the endstop, but it is a small move (presumably at low speed), and it is a compromise between that and potentially dragging across the bed or whatever else is beneath the head.

InsanityAutomation commented:
With advanced pause, we allow either behavior with HOME_BEFORE_FILAMENT_CHANGE as the behavior is nearly identical. For consistency sake, I would prefer to see the option in configuration.h to allow or disallow this behavior. As @sjasonsmith notes, the impact is fairly small however the viable use cases are also few and far between. Personally I would configure my own machines to block the behavior. We could look at the axis trust factor on the function, and possibly allow it if position is is known but not trusted (eg manually set or steppers disabled prior) for a general case finding a middle ground that should be viable on both sides.

classicrocker883 commented:
oh i see so like in configuration.h add something like `#define ALLOW_G27_ZRAISE`, otherwise default to having it trusted.
the only other possible "compromise" is having it only allow to raise Z if axis is !trusted maybe 1 or 2 times.

InsanityAutomation commented:
Yup, adding the config option as described to allow this behavior is the route I would go.

classicrocker883 committed:
Update G27.cpp, add G27_BYPASS_TRUST

classicrocker883 committed:
Update Configuration.h w/ G27_BYPASS_TRUST under NOZZLE_PARK_FEATURE

classicrocker883 commented:
added if `G27_BYPASS_TRUST` is enabled, allows G27 P3 !trusted/homed Z axis
added to Configuration.h under `NOZZLE_PARK_FEATURE`
```y
#define G27_BYPASS_TRUST              // Allows `G27 P3` to raise Z even if axis is not trusted
```

thisiskeithb committed:
Disabled by default

thisiskeithb commented:
> added if `G27_BYPASS_TRUST` is enabled, allows G27 P3 !trusted/homed Z axis
> added to Configuration.h under `NOZZLE_PARK_FEATURE`

https://github.com/MarlinFirmware/Marlin/pull/26789/commits/b47a4457776d7be7a4bf57b4daff8276dffe76ff disables this by default to retain existing behavior.

classicrocker883 committed:
add to sanitycheck

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Update SanityCheck.h

classicrocker883 committed:
Update SanityCheck.h

classicrocker883 committed:
Update SanityCheck.h rearrange

classicrocker883 committed:
Update SanityCheck.h

classicrocker883 committed:
revert SanityCheck.h

classicrocker883 committed:
Update G27.cpp comments

classicrocker883 commented:
I don't understand, what is causing the error in checks? only says operation cancelled on very last one. surely it cannot be something in this **PR** two files

thisiskeithb commented:
> I don't understand, what is causing the error in checks? only says operation cancelled on very last one. surely it cannot be something in this **PR** two files

Sometimes there are issues on GitHub's end. I restarted the failed job(s), so watch the checks page for details: https://github.com/MarlinFirmware/Marlin/pull/26789/checks

classicrocker883 commented:
> Sometimes there are issues on GitHub's end. I restarted the failed job(s), so watch the checks page for details: https://github.com/MarlinFirmware/Marlin/pull/26789/checks

ok thanks, I figured it was something like that. seems to be working. it was strange because it was the same error a couple times in a row

classicrocker883 committed:
Merge https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
use `byteval`

classicrocker883 committed:
Refactor G27 function comments

Updated comments for clarity and consistency in G27 function.

classicrocker883 committed:
Rename to Park Nozzle

classicrocker883 committed:
Update G27.cpp comments

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Feb
